File: year2019/day05/java/Solution.java
import java.nio.file.*;
import java.util.*;

public class Solution {
  private static int[] NUMBERS;

  public static void main(String[] args) {
    try {
      setup(Files.readString(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

	public static int part1() {
    return calc(1);
  }

  public static int part2() {
    return calc(5);
  }

  public static void setup(String input) {
    NUMBERS = Arrays.stream(input.replace("\n", "").split(",")).mapToInt(Integer::parseInt).toArray();
  }

  private static int calc(int mode) {
    int[] program = NUMBERS.clone();
    int result = -1;
    int i = 0;
    while (program[i] % 100 != 99) {
      if (program[i] % 100 == 1 || program[i] % 100 == 2 || program[i] % 100 == 7 || program[i] % 100 == 8) {
        int p1 = program[i + 1];
        int p2 = program[i + 2];
        int p3 = program[i + 3];
        int val1 = getValue(p1, program, (program[i] / 100) % 10);
        int val2 = getValue(p2, program, (program[i] / 1000) % 10);
        if (program[i] % 100 == 1) program[p3] = val1 + val2;
        else if (program[i] % 100 == 2) program[p3] = val1 * val2;
        else if (program[i] % 100 == 7) program[p3] = val1 < val2 ? 1 : 0;
        else if (program[i] % 100 == 8) program[p3] = val1 == val2 ? 1 : 0;
        i += 4;
      } else if (program[i] % 100 == 5 || program[i] % 100 == 6) {
        int p1 = program[i + 1];
        int p2 = program[i + 2];
        int val1 = getValue(p1, program, (program[i] / 100) % 10);
        int val2 = getValue(p2, program, (program[i] / 1000) % 10);
        if ((program[i] % 100 == 5 && val1 != 0) || (program[i] % 100 == 6 && val1 == 0)) {
          i = val2;
        } else {
          i += 3;
        }
      } else if (program[i] % 100 == 3) {
        program[program[i + 1]] = mode;
        i += 2;
      } else if (program[i] % 100 == 4) {
        result = getValue(program[i + 1], program, (program[i] / 100) % 10);
        i += 2;
      }
    }
    return result;
  }

  private static int getValue(int param, int[] program, int mode) {
    return mode == 1 ? param : program[param];
  }
}

File: year2019/day05/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  return calc(input, 1)


def part_2():
  return calc(input, 5)


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [int(num) for num in file.read().split(',')]


def calc(input, mode):
  program = list(input)
  result, i = -1, 0

  def get_value(param, mode):
    return param if mode == 1 else program[param]

  while program[i] % 100 != 99:
    if program[i] % 100 in [1, 2, 7, 8]:
      p1, p2, p3 = program[i + 1:i + 4]
      val1 = get_value(p1, (program[i] - program[i] % 100) // 100 % 10)
      val2 = get_value(p2, (program[i] - program[i] % 100) // 1000 % 10)

      if program[i] % 100 == 1:
        program[p3] = val1 + val2
      elif program[i] % 100 == 2:
        program[p3] = val1 * val2
      elif program[i] % 100 == 7:
        program[p3] = 1 if val1 < val2 else 0
      elif program[i] % 100 == 8:
        program[p3] = 1 if val1 == val2 else 0

      i += 4

    elif program[i] % 100 in [5, 6]:
      p1, p2 = program[i + 1:i + 3]
      val1 = get_value(p1, (program[i] - program[i] % 100) // 100 % 10)
      val2 = get_value(p2, (program[i] - program[i] % 100) // 1000 % 10)

      if (program[i] % 100 == 5 and val1 != 0) or (program[i] % 100 == 6 and val1 == 0):
        i = val2
      else:
        i += 3

    elif program[i] % 100 == 3:
      program[program[i + 1]] = mode
      i += 2

    elif program[i] % 100 == 4:
      result = get_value(program[i + 1], (program[i] - program[i] % 100) // 100 % 10)
      i += 2

  return result



if __name__ == '__main__':
  setup('../puzzle.txt')
  main()